azure-iot-device/tests/iothub/pipeline/test_mqtt_topic_iothub.py

```diff
@@ -0,0 +1,1014 @@
+        topic = "$iothub/twin/PATCH/properties/desired/?$version=1"
+        assert mqtt_topic_iothub.is_twin_desired_property_patch_topic(topic)
+
+    @pytest.mark.it("Returns False if the provided topic is not a desired property patch topic")
+    @pytest.mark.parametrize(
+        "topic",
+        [
+            pytest.param("not a topic", id="Not a topic"),
+            pytest.param("$iothub/methods/POST/fake_method/?$rid=1", id="Topic of wrong type"),
+            pytest.param("$iothub/twin/PATCH/properties/dsiered/?$version=1", id="Malformed topic"),
+        ],
+    )
+    def test_is_not_desired_property_patch_topic(self, topic):
+        assert not mqtt_topic_iothub.is_twin_desired_property_patch_topic(topic)
+
+
+@pytest.mark.describe(".get_input_name_from_topic()")
+class TestGetInputNameFromTopic(object):
+    @pytest.mark.it("Returns the input name from an input topic")
+    def test_valid_input_topic(self):
+        topic = "devices/fake_device/modules/fake_module/inputs/fake_input/%24.mid=6b822696-f75a-46f5-8b02-0680db65abf5&%24.to=%2Fdevices%2Ffake_device%2Fmodules%2Ffake_module%2Finputs%2Ffake_input"
+        expected_input_name = "fake_input"
+
+        assert mqtt_topic_iothub.get_input_name_from_topic(topic) == expected_input_name
+
+    @pytest.mark.it("URL decodes the returned input name")
+    def test_url_decodes_value(self):
+        topic = "devices/fake_device/modules/fake_module/inputs/fake%24input/%24.mid=6b822696-f75a-46f5-8b02-0680db65abf5&%24.to=%2Fdevices%2Ffake_device%2Fmodules%2Ffake_module%2Finputs%2Ffake%24input"
+        expected_input_name = "fake$input"
+        assert mqtt_topic_iothub.get_input_name_from_topic(topic) == expected_input_name
+
+    @pytest.mark.it("Raises a ValueError if the provided topic is not an input name topic")
+    @pytest.mark.parametrize(
+        "topic",
+        [
+            pytest.param("not a topic", id="Not a topic"),
+            pytest.param("$iothub/methods/POST/fake_method/?$rid=1", id="Topic of wrong type"),
+            pytest.param(
+                "devices/fake_device/inputs/fake_input/%24.mid=6b822696-f75a-46f5-8b02-0680db65abf5&%24.to=%2Fdevices%2Ffake_device%2Fmodules%2Ffake_module%2Finputs%2Ffake_input",
+                id="Malformed topic",
+            ),
+        ],
+    )
+    def test_invalid_input_topic(self, topic):
+        with pytest.raises(ValueError):
+            mqtt_topic_iothub.get_input_name_from_topic(topic)
+
+
+@pytest.mark.describe(".get_method_name_from_topic()")
+class TestGetMethodNameFromTopic(object):
+    @pytest.mark.it("Returns the method name from a method topic")
+    def test_valid_method_topic(self):
+        topic = "$iothub/methods/POST/fake_method/?$rid=1"
+        expected_method_name = "fake_method"
+
+        assert mqtt_topic_iothub.get_method_name_from_topic(topic) == expected_method_name
+
+    @pytest.mark.it("URL decodes the returned method name")
+    def test_url_decodes_value(self):
```

Comment:
>test_url_decodes_value [](start = 8, length = 22)

this only tests $, do we need to test the others (space, slash, plus)?

Reply:
Not really - it's important we test the weird values on encode, but on decode, things are a lot simpler. The only variance is on the '+' character, but since '+' is impossible in MQTT....... it's not really necessary IMO.

That said I could add it for completeness.

---
In reply to: [399805654](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r399805654) [](ancestors = 399805654)